Automated BIP39 Recovery

lukechilds opened the pull request:
Resolves #6155 

Implements automated scanning of all known BIP32 chains that can be imported into Electrum.

I'm opening this PR for review, it's not ready to be merged yet. Account discovery is functional but still needs to be integrated in to the GUI.

Currently it's implemented as a script for easy testing, you can quickly test it out by running the `electrum/scripts/bip39_recovery.py` script with a `mnemonic` parameter and also an optional second `passphrase` parameter. It will return a list of any previously used accounts it can find. You can use this mnemonic that I've created a few test accounts in:

```
much bottom such hurt hunt welcome cushion erosion pulse admit name deer
```

e.g:

```
$ ./bip39_recovery.py "much bottom such hurt hunt welcome cushion erosion pulse admit name deer"
[
    {
        "derivation_path": "m/84'/0'/0'",
        "description": "Standard BIP84 native segwit (Account 0)",
        "script_type": "p2wpkh"
    },
    {
        "derivation_path": "m/84'/0'/1'",
        "description": "Standard BIP84 native segwit (Account 1)",
        "script_type": "p2wpkh"
    },
    {
        "derivation_path": "m/0'",
        "description": "Non-standard legacy (Account 0)",
        "script_type": "p2pkh"
    },
    {
        "derivation_path": "m/84'/0'/2147483646'",
        "description": "Samourai Whirlpool post-mix",
        "script_type": "p2wpkh"
    }
]
```

**Diversions from BIP44**

Due to the fact that at discovery we only need to check which accounts has been used, not the final balance of each account, we take a few shortcuts:
- Don't scan the change chain, if a BIP44 compliant account is used it will have received a TX in one of the first 20 (BIP44 gap limit) addresses of the external chain.
- Move on to the next account as soon as we find a single address with funds. No need to keep scanning until we exhaust the gap limit.

If one of the discovered accounts is selected by the user for recovery, then Electrum will proceed to properly scan both internal external chains and exhaust the gap limit when it creates the wallet.

**Limitations**

Electrum takes an account level import path like `m/84'/0'/0'` and derives the wallet with the non hardened `is_change/address_index` sub trees. Due to this we cannot import Bitcoin Core accounts as they use the format [`m/0'/0'/address_index'`](https://github.com/bitcoin/bitcoin/issues/13302#issuecomment-391227927), notice all sub trees use hardened derivation.

All other wallets listed on walletsrecovery.org can be successfully recovered.

---

Please let me know if this code looks ok so far. Next steps will be to migrate the account discovery code from the script to `electrum.bip39_recovery` and then integrate it in to the recovery wizard GUI.

CC @ecdsa @SomberNight @aantonop

lukechilds commented on electrum/scripts/bip39_recovery.py:
```diff
@@ -0,0 +1,127 @@
+def format_account(wallet_format, account_path):
+    description = wallet_format["description"]
+    if wallet_format["iterate_accounts"]:
+        account_index = account_path.split("/")[-1].replace("'", "")
```
Good point, will do.

lukechilds replied:
Resolved in https://github.com/spesmilo/electrum/pull/6219/commits/9855fe2115eb341cf359fb128cc84bb242f27f32.

lukechilds commented:
Thanks for the review @SomberNight, I have two questions re GUI integration.

**1. Whats the best way to hook into the derivation path view?**

I want to add a button in this view that says "Detect Existing Accounts" which when clicked, opens a popup with a loading icon, calls the account discovery function, displays a list of results to the user, allows them to select one, and populate the path/script type inputs with the result.

<img width="712" alt="Screenshot 2020-06-08 at 21 08 46" src="https://user-images.githubusercontent.com/2123375/84040096-66b85380-a9cc-11ea-85f1-39c03c8942aa.png">

For example the previous view to enter the seed has it's own file in `electrum/gui/qt/seed_dialog.py`, so I could just edit that. However the derivation path is implemented in a GUI diagnostic way in `electrum/base_wizard.py` which sets up some params and then inits the UI with `self.choice_and_line_dialog()` where `choice_and_line_dialog()` is a generic view provided by the GUI implementation.

https://github.com/spesmilo/electrum/blob/d0ab0039784d1d0faade548182f72aa6d47b7e82/electrum/base_wizard.py#L407-L439

I can see the QT implementation here:

https://github.com/spesmilo/electrum/blob/d0ab0039784d1d0faade548182f72aa6d47b7e82/electrum/gui/qt/installwizard.py#L605-L633

I could just edit that method and wrap my code with. `if title == "Script type and Derivation path"` but that seems like a pretty nasty solution and would also break when i18n is active.

What's the best way to hook into this?

**2. How should I provide network access to my method?**

Currently in the script I'm creating a network instance with:

```python
from electrum.simple_config import SimpleConfig
from electrum.network import Network
config = SimpleConfig()
network = Network(config)
network.start()
```

When I add this to the GUI I'll obviously want to use the existing Network instance. I see `Network` is a singleton, and I can get the instance with `Network.get_instance()`, however I'd also like to keep the script functional, by just updating it to import `electrum.bip39_recovery` and running it. It's useful to quickly test during dev and seems like a pretty handy script to expose to users too.

What's the best way to go about this? Create a helper that returns the global instance if it exists or creates one if it doesn't? Or maybe pass a `network` instance in to the function, and pass in the one I create for the script or pass in the global instance for the GUI?

Not quite sure the best way to go about this.

SomberNight commented:
> 2. How should I provide network access to my method?
> Or maybe pass a network instance in to the function, and pass in the one I create for the script or pass in the global instance for the GUI?

Yes, that sounds good.

> GUI I'll obviously want to use the existing network instance (if it exists during account creation?)

Yes, typically it exists, unless the user used the `--offline` CLI flag.
When you get a reference to network, you should test if it's None, and if it is, just display an error message that they are offline.

> 1. Whats the best way to hook into the derivation path view?
> the derivation path is implemented in a GUI agnostic way in electrum/base_wizard.py which sets up some params and then inits the UI with self.choice_and_line_dialog() where choice_and_line_dialog() is a generic view provided by the GUI implementation.

`choice_and_line_dialog` is already only used for this specific usecase.
I think you can just rename it to something "derivation"-specific, and modify it directly.
I imagine you only want to implement all this for the Qt GUI? (which is fine)
Make sure that the change does not break the kivy GUI.
i.e. rename the method, pass it whatever arguments you need, make sure both GUIs accept all the arguments even if they ignore some, make sure the return types are the same
Then you can just implement extra logic in the Qt-specific implementation of this method -- well you probably want to implement it somewhere else as I imagine it would be many lines of code, and just call it from there.

lukechilds commented:
@SomberNight thanks, makes sense!

Hadn't realised `choice_and_line_dialog` was only used for this one dialog.

And yes, I was only planning on implementing it in the QT GUI. CLI users can use the script, and I don't think the Kivy GUI allows setting the BIP39 option on a seed so not really applicable there.

lukechilds commented:
@SomberNight when calling an async function from a QT `WindowModalDialog` I can do:

```python
accounts = asyncio.run_coroutine_threadsafe(account_discovery(), network.asyncio_loop).result()
```

Which works but it blocks the QT thread so the UI hangs and there's no way for the user click the cancel button while they're waiting.

I'd like to:

- Show the modal
- Show some text like "Scanning for existing accounts..." or maybe an animated spinner
- Fire off the account discovery to run asynchronously
- Allow the user to interact with the UI, cancel the modal if they want
- Once account discovery has completed, update the modal with the results

Are you able to point me in the right direction for how to achieve this?

**Disclaimer:** Apologies if this question seems obvious, I'm not really a Python developer, in fact the only other time I've ever written Python was for my SSL fingerprint PR. I'm not familiar with Python's async API works.

SomberNight commented:
> if this question seems obvious

No worries, it's not obvious at all. Qt does not really support python asyncio.

Take a look at https://github.com/spesmilo/electrum/blob/b0230f6a4bd7498aa0dd1d48ec61326d35a40386/electrum/gui/qt/main_window.py#L1658

It does not behave exactly as you described but similar.

lukechilds commented:
I've integrated the functionality into the GUI.

Demo:

![electrum_bip39](https://user-images.githubusercontent.com/2123375/84978393-f9888900-b156-11ea-9ee4-fd77246a55a4.gif)

rdymac commented:
The resulting file name could use another label (as the [standard] one), so the user knows which one he selected after the scan.

jlopp commented:
I'd highly encourage using a gap limit that far exceeds the standard. Consider the fact that there are a variety of applications such as payment processing that can result in huge gaps. If I recall correctly, btcpayserver set their gap limit to either 1,000 or 10,000. In my opinion the additional computation required is well worth the pain it will save for folks who have funds stranded beyond the standard gap limit.

Enegnei commented:
> I'd highly encourage using a gap limit that far exceeds the standard. Consider the fact that there are a variety of applications such as payment processing that can result in huge gaps. If I recall correctly, btcpayserver set their gap limit to either 1,000 or 10,000. In my opinion the additional computation required is well worth the pain it will save for folks who have funds stranded beyond the standard gap limit.

For [reference](https://github.com/btcpayserver/btcpayserver/blob/3b91b380146182d48d0e41262075e8f141ac9249/BTCPayServer/Models/WalletViewModels/RescanWalletModel.cs), yes, their gap limit appears to be 10,000.

SomberNight commented:
> I'd highly encourage using a gap limit that far exceeds the standard.

Do you mean for the discovery, or after the wallet is created for normal operations?

If you mean for the discovery, then ok, perhaps. It seems a bit unlucky to have a wallet where none of the first few addresses are used but later ones are, but I guess it can happen if a merchant was griefed as soon as they started using a wallet.

But to be clear, regardless of what insane gap limits are some services are using, we are not able to use insane high gap limits simply due to architecture limitations. IMO what these services do when they use a gap limit of 10000 by default is anti-social behaviour: they are pulling the whole ecosystem in the same direction and slowly forcing everyone to use larger and larger gap limits (what if poor user used insane service no.17, we should use a higher gap limit then...). By using those high gap limits they introduce the risk for the user that they might not be able to restore easily using other services. That's just how it goes, we are not participating in this game.

lukechilds commented:
Some back-of-the-envelope calculations re gap limit:

It currently takes around 2 seconds per account for discovery to complete with a gap limit of 20.

That means in the best case scenario where there are 0 used accounts, we will only scan 13 paths so it'll take ~26 seconds.

Setting the gap limit to 10,000 will mean the **best case** recovery process increases from ~26 seconds to ~3.6 hours.

SomberNight commented:
> Setting the gap limit to 10,000 will mean the best case recovery process increases from ~26 seconds to ~3.6 hours.

Default ElectrumX DOS limits will get hit a lot sooner than that, slowing you down to oblivion.

lukechilds commented:
One solution could be to read the gap limit from the Electrum config if it exists instead of hardcoding to 20. Or maybe add a new `BIP39_recovery_gap_limit` param specifically for this.

Then if you connect to your own local Electrum instance, disable DoS limits, and manually set the config param to 10,000 it would work.

But if you're capable of all this, you're probably also capable of just manually restoring your wallet.

aantonop commented:
The audience for this "wizard" is newbies who don't know their derivation
path. The standard gap limit on import is enough. This is not a tool for
merchants or advanced users.

Let's keep it as simple and efficient as possible.

lukechilds commented:
>The resulting file name could use another label (as the [standard] one), so the user knows which one he selected after the scan.

@rdymac that would indeed be nice but unfortunately that isn't really possible without making quite large changes to the way the Electrum import wizard works.

Currently this PR is just adding a modal to one window and then auto-filling some fields in that window on completion.

Keeping this change simple and not too intrusive into the way Electrum works was a specific goal for this PR: https://github.com/spesmilo/electrum/issues/6155#issuecomment-629783062

SomberNight commented:
>> The resulting file name could use another label (as the [standard] one), so the user knows which one he selected after the scan.
>
> @rdymac that would indeed be nice but unfortunately that isn't really possible without making quite large changes to the way the Electrum import wizard works.

Yes, let's not change `[standard]`, it is related to the wallet type.
Note that the chosen script type is visible in `Wallet`>`Information`.
The derivation path is not visible in the GUI atm, but we could add it to the same dialog. If you want this, please see https://github.com/spesmilo/electrum/issues/4700 (and move that discussion there).

ecdsa commented:
> IMO what these services do when they use a gap limit of 10000 by default is anti-social behaviour

that's my opinion too. a high limit might work if you are scanning blocks yourself, but we should not engage in such behaviour with public servers.

SomberNight commented on electrum/bip39_recovery.py:
```diff
@@ -0,0 +1,51 @@
+        scripthash = bitcoin.script_to_scripthash(script)
+        history = await network.get_history_for_scripthash(scripthash)
+        if len(history) > 0:
+            return True
```
This `for` [...] `await task` construct here will mean that we send a request to the server, wait for it to answer, and then send the next request. This is why it is taking as long as you said (several seconds or similar).
That is what the `await` does, it actually stops the coroutine, and asks to only be rescheduled after that task has returned (mid-loop).

Instead, I suggest you use a TaskGroup. A TaskGroup lets you create a context (a `with` block) which will only exit once all the tasks inside returned; or if any task raises, it will cancel the remaining tasks and just re-raise the exception (so that it can propagate out).
That way, you can send all the requests "concurrently", i.e. send them all, and handle the responses as they arrive.
See these examples:
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/interface.py#L567
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/interface.py#L594
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/tests/test_lnpeer.py#L504

Note that atm these nested loops you have will send all N*20 requests in series, so we will wait for over 100 network RTTs. By putting a TaskGroup here, you can reduce that to around N RTTs (N being the number of derivation path prefixes).
As further optimisation, you could use a TaskGroup at the outer loop too, so that it becomes just a single RTT (disregarding the account stuff for the moment).

lukechilds replied:
Ahh, thanks!

I actually attempted to get all requests going concurrently by first saving executing the async functoins and saving the return values in a variable and then awaiting them later.

Something like:

```python
requests = []
for url in urls:
  requests.append(fetch(url))
results = []
for request in requests:
  results.append(await request)
```

This works in JavaScript since the async function is triggered by the function execution, `await` is just used to wait for the response.

Seems like in Python the async function execution just returns some "thing" that represents an async function call and doesn't actually execute anything until it is awaited, is that correct?

Since the above code still seemed to execute the requests serially.

e.g, equiv JavaScript code is:

```js
equests = []
for (url of urls) {
    requests.push(fetch(url))
}
results = []
for (request of requests) {
    results.push(await request)
}
```

All `fetch()` requests actually happen concurrently here and once the slowest request has completed you and up with `results` being an array of all the resolved values.

I'll have a play with `TaskGroup` and try and get it working concurrently.

Also, is there not a risk we hit ElectrumX rate limits if we do all requests concurrently?

lukechilds replied:
I implemented concurrency, both at the address lookup level and the derivation scheme level.

The entire discovery process now takes ~2s for my test seed. 🎉

lukechilds commented on electrum/base_wizard.py:
```diff
@@ -492,7 +492,11 @@ def restore_from_seed(self):
-            f = lambda passphrase: self.on_restore_bip39(seed, passphrase)
+            def f(passphrase):
+                # TODO: Check with @SomberNight if this is ok
+                self.seed = seed
```
@SomberNight is it ok to persist sensitive data on the `BaseWizard` instance like this?

I'm attaching `seed` and `passphrase` to the instance so I can access them later in the derivation dialog.

I'm not exactly sure how long the `BaseWizard` instance lives for, but I assume it's destroyed once the wizard has completed so `seed`/`mnemonic` would be garbage collected.

SomberNight replied:
As on IRC, I think (seed, passphrase) could be passed around as parameters. Probably optional keyword-only args, as the hardware wallet codepath will not have them.

lukechilds commented:
I've addresses the feedback and tested/fixed when setting up with a HWW wallet.

Unless storing seed data on `BaseWizard` is an issue https://github.com/spesmilo/electrum/pull/6219#discussion_r443206598 I think this is ready to be merged.

Let me know if you require any other changes or you think there's any other scenarios I should test.

SomberNight commented:
To expand on https://github.com/spesmilo/electrum/pull/6219/files#r444897453 ,
I think what we could pass around instead is a function `(derivation_prefix) -> (account_xpub)` (or similar), as then this could be (later?) implemented for hardware wallets as well.

andronoob commented:
@aantonop

>This is not a tool for merchants or advanced users.

Why it can't? In my opinion, it can bring convenience for them as well. I think this may be implemented in a progressive favour, with some rate limit to avoid unintentionally DoSing the server.

andronoob commented:
I also wonder whether the "BIP39-Electrum2.0 duality" issue (#1300) is properly handled here?

Electrum 2.0 seed phrase is not compatible with BIP39, but it still shares exactly the same wordlist with BIP39. Therefore, an Electrum 2.0 seed can be both a valid Electrum 2.0 seed and a valid BIP39 mnemonic, with a probability of one in sixteen. (12-word BIP39 mnemonic phrase has 4-bit checksum. 2^4=16)

It's then possible (although not very possible, because the BIP39 wallet would show empty transaction history and zero balance, so that the user is highly unlikely to continue) for a user to import an Electrum 2.0 seed into another BIP39-compliant wallet.

andronoob commented:
I dreamed that the new derivation path step of the wizard may look like this:


### Specify your derivation path
----
#### Standard derivation paths
- Legacy (P2PKH)
  Example: 1CuC3b...
- SegWit, compatible (P2SH-P2WPKH)
  Example: 3Dq4Y9...
- SegWit, native Bech32 (P2WPKH)
  Example: bc1qff6a...

  Account number: _____
#### Non-standard derivation paths
- `Click here to select from the list...`
#### Not sure?
- `Let me guess!`

**(empty, if no detection is ever executed) / No fund detected / Recoverable funds detected, congratulations! You may click "next" now.** `[Wait for 30s] Detect funds / (Button hidden if funds detected) / Click to probe the next XXXX addresses`

----
  *Don't touch things below if you have no idea what they mean.*

  BIP32 derivation path: _________  Address type: `P2WPKH (dropdown menu)`

  Gap limit: ____
- [ ] Manually specify the starting child key index: ____

ecdsa commented:
related: #6001

lukechilds commented:
@SomberNight @ecdsa are you able to re-review when you have time?

I've resolved the issue of attaching the mnemonic/passphrase to the instance and instead now pass an `(account_path) -> (account_xpub)` function around.

I haven't implemented this function for when a HWW is being restored but I have tested the wizard and it correctly leaves out the restore option when in HWW mode.

lukechilds commented:
@SomberNight @ecdsa sorry to keep pinging you guys, it's been a month now since the last time I checked in with you.

AFAIK I've implemented all required changes and there aren't any conflicts, is there anything else you want from me for this or can we get it merged in?

SomberNight commented:
Starting with https://github.com/spesmilo/electrum/commit/41827cff1943819c221acf2e2be675d049b4b7c4 (copy here: https://github.com/SomberNight/electrum/tree/pr/6219_bak20200820),
I've squashed all your commits, https://github.com/spesmilo/electrum/commit/cc7a6e8e64585d241c6ad2fd85a38c420a81bf3f,
and rebased on master, https://github.com/spesmilo/electrum/commit/7b122d26792af7e856c8e841dff8f4c5e83c3566.

SomberNight commented on electrum/bip39_recovery.py:
```diff
@@ -0,0 +1,75 @@
+        "description": description,
+        "derivation_path": bip32_ints_to_str(account_path),
+        "script_type": wallet_format["script_type"],
+    }
```
Would be better to return a NamedTuple or similar; e.g. so that fields can be typed and can be enumerated statically (by IDE).

SomberNight reviewed (approved):
Looks good. Reviewed code and tested.

I have some minor nits (in addition to what I pushed) but that's fine.
Noticed that the Qt dialog has a sizing bug on dark theme (too small by default) - might be platform dependent, and probably related to QDarkStyle, so ignoring for now (my naive attempts to fix it failed).

SomberNight commented:
Thank you for your work.

AlistairTB commented:
I would like to do exactly this with a multisig wallet using Zpub keys (for a read-only wallet). In case it's not obvious, the Zpubs were derived from BIP39 words and extended passphrase. Is this possible?

> I've integrated the functionality into the GUI.
> 
> Demo:
> 
> ![electrum_bip39](https://user-images.githubusercontent.com/2123375/84978393-f9888900-b156-11ea-9ee4-fd77246a55a4.gif)
> 
> Let me know if it looks ok or if you have any suggestions, thanks!

SomberNight commented:
@AlistairTB No, this is only implemented for singlesig wallets.

Also, it is not clear exactly how you would want to use this.
The functionality here is for bruteforcing the derivation path and script type.
For a multisig wallet given Zpubs, you already know the script type, and you would typically have the Zpubs already at the "account level" of the derivation, so you already know the derivation path.
Further note that it is not possible to do hardened derivation starting from a master public key.

AlistairTB commented:
Thank you for the reality check. I was using it the way it was intended, however, I had the wallet type wrong (3/3 vs 2/3) and was looking to solve the wrong problem.

SomberNight commented:
> I'd say this feature would be very well appreciated also when creating a watch-only wallet from a hardware device. Shall I open a dedicated issue?

Sure, go ahead.
